### PR TITLE
Add and edit users over the API

### PR DESCRIPTION
**Pull request by huwshimi**
Done:
- Add sagas/actions/reducers for creating/updating users.
- Connect the user create/update forms to the API. Fixes: https://github.com/canonical-web-and-design/maas-squad/issues/1373.

QA:
- Navigate to the users list.
- Click on the Add user button.
- Fill out the form and submit.
- You should be redirected to the user list and the new user should be in the list.
- Repeat with editing a user.
Note: a message should appear when you are redirected: https://github.com/canonical-web-and-design/maas-ui/issues/108.

### CODE REVIEW

File: ui/src/app/settings/utils.js [hunk 1]
```diff
@@ -20,16 +20,14 @@ const simpleObjectEquality = (obj1, obj2) => {
...
-  const { initialValues, isSubmitting, errors, values } = formikProps;
+  const { initialValues, errors, values } = formikProps;
   let hasErrors = false;
 
   if (errors) {
     hasErrors = Object.keys(errors).length > 0;
   }
 
-  return (
-    isSubmitting || simpleObjectEquality(initialValues, values) || hasErrors
```
Comment (huwshimi) [inline ui/src/app/settings/utils.js]: isSubmitting causes issues when controlling the state from redux (e.g. when server errors are returned in the state).

File: ui/src/app/base/components/FormikField/FormikField.js [hunk 2]
```diff
@@ -12,15 +12,29 @@ const FormikField = ({
...
+    touched
+  } = formikProps;
+  const { serverErrors = {}, invalidValues = {} } = status;
   return (
     <Component
-      error={formikProps.touched[fieldKey] && formikProps.errors[fieldKey]}
+      error={
+        (touched[fieldKey] &&
+          (errors[fieldKey] ||
+            (value === invalidValues[fieldKey] && serverErrors[fieldKey]))) ||
```
Comment (Caleb-Ellis) [inline ui/src/app/base/components/FormikField/FormikField.js]: This conditional is a bit hard to parse. Is it possible to split it out into something readable like:
``` javascript
error={isTouched && (hasFormikErrors || hasServerErrors)}
```
?

File: ui/src/app/settings/utils.js (hunk 1 above)
Comment (Caleb-Ellis) [reply ui/src/app/settings/utils.js]: That's fair. Might be a good idea to disable the button if `saving` is true in that case.

Comment (huwshimi) [reply ui/src/app/settings/utils.js]: Is there a way to disable the button without the vanilla styles changing? Maybe it's OK for them to change, would just look different to the designs that's all.

Comment (huwshimi) [reply ui/src/app/settings/utils.js]: Nevermind, looks like that case is already handled in the action button.

File: ui/src/app/base/components/FormikField/FormikField.js (hunk 2 above)
Comment (huwshimi) [reply ui/src/app/base/components/FormikField/FormikField.js]: Thanks, updated.

File: ui/src/app/settings/views/Users/UserFormFields/UserFormFields.js [hunk 3]
```diff
@@ -15,6 +19,22 @@ const togglePassword = (event, passwordVisible, showPassword) => {
...
+    if (Object.keys(errors).length) {
+      const formikErrors = {};
+      const invalidValues = {};
+      Object.keys(errors).forEach(field => {
+        formikErrors[field] = errors[field].join(" ");
+        invalidValues[field] = formikProps.values[field];
+      });
+      formikProps.setStatus({ serverErrors: formikErrors, invalidValues });
+    }
+  }, [errors]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (squidsoup) [inline ui/src/app/settings/views/Users/UserFormFields/UserFormFields.js]: I think we should revisit this, but I have a migraine, so now is not the time.

File: ui/src/app/settings/reducers/users.js [hunk 4]
```diff
@@ -11,14 +11,49 @@ const users = produce(
...
+      case "CREATE_USERS_SUCCESS":
+        draft.errors = {};
+        draft.saved = true;
+        draft.saving = false;
+        break;
+      case "UPDATE_USER_SYNC":
+        for (let i in draft.items) {
+          if (draft.items[i].id === action.payload.id) {
+            draft.items[i] = action.payload;
+            break;
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/settings/reducers/users.js]: I've updated the config reducer to match this one. This one is more performant as it breaks once it finds the item to update instead of mapping over all items.
